### ISSUE TITLE
Generated memes do not appear in iOS app

### ISSUE THREAD

**Issue by hmenzagh** [mattermost/mattermost-plugin-memes#41]
Everything is in the title, just did a fresh install (Mattermost & `mattermost-plugin-memes`) works like a charm on desktop & web, but I get blank posts from the mobile version.

**Comment by hanzei** [comment]
Do you see any errors in the server logs?

**Comment by hmenzagh** [comment]
Nope, I haven't got anything, unless there is a config to change to see `mattermost-plugin-memes`'s logs

**Comment by kevinbosman** [comment]
I'm seeing the same problem on Android Mattermost. Memes with standard text appear fine, but if there's custom text it does not show any meme at all. Nothing in the logs unless you use a single quote in the text, then logs say "Unterminated single-quoted string"

Works perfectly on desktop and web.

**Comment by Twilek-de** [comment]
Seeing the same problem. Memes show up in Web and in Desktop Client, but not in IOS (stable and beta) and Android clients..

**Comment by moschlar** [comment]
Here too…

**Comment by nathanaelhoun** [comment]
Same here. memes v1.4.0 compiled on my machine. Server is 5.38.1, mobile is v1.46. 
No logs in Mattermost server related to the plugin.

- The image correctly loads if I go to the link in my mobile browser. 
- The image correctly loads if I replace MY_MATTERMOST_URL by the community.mattermost.com.
- The image doesn't load if I copy-paste the post in my mobile app connected to community.mattermost.com.
- The image doesn't load with a local mattermost server and a mobile app connected to this local server, with ServerURL correctly configured (so I think it is not a proxy issue)
```
![am-i-the-only-one](https://MY_MATTERMOST_URL/plugins/memes/templates/am-i-the-only-one.jpg?text=hello&text=you)
```

**Comment by nathanaelhoun** [comment]
Some more troubleshooting informations: when querying the Mattermost API from my instance, the `metadata` field is empty, whereas in Mattermost Community, we have metadata for the image. For what I have seen in the React Native code, it's this metadata field which is used to render images.

<details><summary>Community.mattermost.com API response</summary>

```json
{
  "id": "REDACTED",
  "create_at": 1633522416109,
  "update_at": 1633525971148,
  "edit_at": 1633525971148,
  "delete_at": 0,
  "is_pinned": false,
  "user_id": "REDACTED",
  "channel_id": "REDACTED",
  "root_id": "",
  "original_id": "",
  "message": "![ancient-aliens](https://community.mattermost.com/plugins/memes/templates/ancient-aliens.jpg?text=salut-2)",
  "type": "",
  "props": {
    "override_icon_url": ""
  },
  "hashtags": "",
  "pending_post_id": "",
  "reply_count": 0,
  "last_reply_at": 0,
  "participants": null,
  "metadata": {
    "images": {
      "https://community.mattermost.com/plugins/memes/templates/ancient-aliens.jpg?text=salut-2": {
        "width": 500,
        "height": 436,
        "format": "jpeg",
        "frame_count": 0
      }
    }
  }
}

```
</details>

**Comment by nathanaelhoun** [comment]
Hi @shazm, thanks for digging into this!
I tried the command `/meme ancient-aliens "Aliens man!"`, a variation with several arguments `/meme ancient-aliens "Aliens" "man!"` and without argument `/meme ancient-aliens`.
In each case, both sent from mobile (android) or web :
- the command doesn't fail (no error message, no server log, the message is sent)
- On web, the meme shows correctly with the good text 
- On Android, there is an "empty" message, without any text or image showing, just a small empty space.

Hope it helps!

**Comment by vigstudio** [comment]
This bug exists on both iOS and Android.

**Comment by omginput** [comment]
This bug exists now for two years?

**Comment by theme1256** [comment]
I think I've found the issue... And it's a stupid one...

## tl:dr;
The mobile-app does not support markdown images!

## The full story
After a somewhat short debugging session I discovered that this plugin is sending the images as markdown.  
First I though that it might be due to multiple domains on the same server (havn't ruled that one out yet), but if I manually send a message with markdown, that's supposed to display an image, it doesn't work in the mobile-app, but works fine on desktop (web and app).  
And it is not due too total lack of markdown support in the mobile-app, because headers, list and hightlighting works, just images that's not working.
Don't know if this means that we have to submit the bug elsewhere....

**Comment by khuongphp** [comment]
Sorry, anyone try this? Is issue fixed?

**Comment by knivram** [comment]
I did some digging and I'm pretty sure I found the issue.
## Root Cause

When the memes plugin creates a post, it embeds the meme as a markdown image pointing back to the server itself:
```
![meme-name](siteURL/plugins/memes/templates/meme.jpg?text=...)
```

The Mattermost server then tries to fetch this URL to populate `imagesMetadata` (dimensions, format) for clients. But the HTTP client used for this has SSRF protection that blocks requests resolving to the server's own IP — via [`IsOwnIP()` and `IsReservedIP()`](https://github.com/mattermost/mattermost/blob/24d3fed77757dcc621deef5a92d2ec932fad4c4d/server/public/shared/httpservice/httpservice.go#L88).

There is a bypass for same-host requests, but it only covers the `/api/v4/image` proxy endpoint — plugin URLs at `/plugins/...` are not included.

**Result:** posts from the memes plugin arrive at clients with empty `imagesMetadata`.
## Impact on Mobile

The mobile app's markdown renderer **refuses to render images entirely** when `imagesMetadata` is missing:
```typescript
const renderImage = useCallback((...) => {
    if (!imagesMetadata || isUnsafeLinksPost) {
        return null;
    }
    ...
}
```
This is the [`renderImage` function in `markdown.tsx`](https://github.com/mattermost/mattermost-mobile/blob/1cea8c6f0c6ee484becbefc85ef08379b8cbc740/app/components/markdown/markdown.tsx#L429). No metadata means the image is simply not shown.
## Proposed Fix

There are two ways to tackle this:
1. Update the mobile app to handle missing `imagesMetadata` gracefully (treats the symptom)
2. Fix the plugin so the server can actually load the image metadata (fixes the root cause)

I'd go with option 2: change the plugin to render the meme in-process, upload it via `p.API.UploadFile()`, and attach it to the post via `FileIds`. This way:
- The image is stored as a proper file attachment with full metadata
- The image metadata is fetched through Mattermost's own image endpoint, which is excluded from the SSRF restriction
- It works on all clients without any client-side changes

I have a working implementation of this approach and did a PR to my own fork to showcase it: https://github.com/knivram/mattermost-plugin-memes/pull/1
